olm/operator-registry.md

```diff
@@ -0,0 +1,324 @@
+
+Acceptance Criteria:
+
+When I install an operator index onto my cluster, I can use the data that is available on the cluster to drive operator installation and lifecycle.
+
+---------------------
+
+_**Polling Updates**_
+
+As an OpenShift administrator, I want a way to configure a poll timer to ensure my index of operators is up to date and references all of the latest operator bundles.
+
+- So that I can get updates over time
+- So that I can know the frequency that those updates will appear
+
+Acceptance Criteria:
+When I set the update frequency on an operator index, OLM will poll and ensure that if there is a new version of the index that those changes are pulled and propagated into the cluster.
+
+Internal details:
+Polling process should be reviewed by quay team to ensure we scale on quay (although these indexes should work on ANY container image registry)
+
+### Implementation Details
+
+The goal of this enhancement is to replace the app-registry implementation in a way that is usable by operator developers and CI pipeline developers. To do this we will build an index image that is defined from a set of operator bundle images that are released independently and optimized in a way that can be productized by operator authors and pipeline developers.
+
+This implementation makes the assumption that a separate process is built that generates container images that contain individual operator bundles for each version of an operator. The intent of the index image is to aggregate these images that are out of scope of this enhancement. Once that image is built, the image itself is immutable. In order to create a new image, the tooling described here will run off cluster. Once the image is loaded onto the cluster, the only way to load new content is to pull down a new version of that image (either manually or by an automated poll).
+
+The final implementation goal of this enhancement involves this set of steps:
+
+1. Use a reference to a container image (also referred to here as the operator index) rather than app-registry namespaces to expose installable operators on a cluster
+2. Build an operator-registry database that serves the data required by OLM to drive install and UI workflows
+3. Have a way to optimize the build workflow of operator-registry databases (which currently drive OLM's workflows) from previous versions of the database plus new content (ex. new operator at new version) so that the database need not be built from scratch every time they are created.
+
+To start, let's define what the operator-registry does today. It:
+
+1. Takes a set of manifest files and outputs a database file.
+2. Takes a database file and serves a grpc api that serves content from that db
+3. Takes a configmap reference, builds a db and serves a grpc api that serves content from that db
+4. Takes an app-registry path/namespace reference, builds a db and serves a grpc api that serves content from that db
```

Comment:
3. and 4. are for specific variants of operator-registry, while 1 and 2 are abstract and apply to all variants. The way this is currently written, I'm expecting them to be a set of successive steps rather than features.

Reply:
So, to be clear, these are all directly translated from the four CLI binaries built from the operator-registry today:

1. `initializer`
2. `registry-server`
3. `configmap-server`
4. `appregistry-server`

Agreed that 3 and 4 are specific implementations of 2, but that is the distinction that exists in terms of today's tooling.